addons/html_builder/static/src/core/builder_selection_restriction_plugin.js

```diff
@@ -0,0 +1,242 @@
+            );
+            const closestDiv = closestElement(editableSelection.anchorNode, "div");
+
+            // if the closest special block doesn't contains the closest div block,
+            // we select the paragraph, otherwise we select the whole closest div block
+            if (closestSpecialBlock && !closestSpecialBlock.contains(closestDiv)) {
+                this.selectAllInElement(closestElement(editableSelection.anchorNode, "p"));
+            } else {
+                this.selectAllInElement(closestDiv);
+            }
+            ev.stopPropagation();
+        }
+    }
+
+    // we extend the selection to the whole element step by step to properly
+    // handle uncrossable elements (like row, blockquote...)
+    selectAllInElement(element) {
+        let selection = this.dependencies.selection.getEditableSelection();
+        let { anchorNode, anchorOffset, focusNode, focusOffset, direction } = selection;
+
+        const [newAnchorNode, newAnchorOffset] = getDeepestPosition(element, 0);
+        const [newFocusNode, newFocusOffset] = getDeepestPosition(element, nodeSize(element));
+        if (direction === DIRECTIONS.RIGHT) {
+            this.dependencies.selection.setSelection({
+                anchorNode: focusNode,
+                anchorOffset: focusOffset,
+                focusNode: newAnchorNode,
+                focusOffset: newAnchorOffset,
+            });
+        } else {
+            this.dependencies.selection.setSelection({
+                anchorNode: anchorNode,
+                anchorOffset: anchorOffset,
+                focusNode: newAnchorNode,
+                focusOffset: newAnchorOffset,
+            });
+        }
+        this.correctSelectionOnUncrossable();
+        selection = this.dependencies.selection.getEditableSelection();
+        ({ focusNode, focusOffset } = selection);
+        this.dependencies.selection.setSelection({
+            anchorNode: focusNode,
+            anchorOffset: focusOffset,
+            focusNode: newFocusNode,
+            focusOffset: newFocusOffset,
+        });
+        this.correctSelectionOnUncrossable();
+    }
+
+    restrictSelectionInElement(selection, block) {
+        const { anchorNode, anchorOffset, focusNode, direction } = selection;
+        const isFocusInBlock = block.contains(focusNode);
+
+        if (!isFocusInBlock) {
+            let focusNode, focusOffset;
+            if (direction === DIRECTIONS.RIGHT) {
+                [focusNode, focusOffset] = getDeepestPosition(block, nodeSize(block));
+            } else {
+                [focusNode, focusOffset] = getDeepestPosition(block, 0);
+            }
+            this.dependencies.selection.setSelection({
+                anchorNode,
+                anchorOffset,
+                focusNode,
+                focusOffset,
+            });
+        }
+        this.correctSelectionOnUncrossable();
+    }
+
+    restrictSelectionInClosestDiv(ev) {
```

Comment:
- doc string
- technically, the name is not right, since it can be something else than a `div`